modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/IndexCorruptionRebuildTest.java

```diff
@@ -0,0 +1,347 @@
+        ));
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        stopAllGrids();
+
+        cleanPersistenceDir();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+
+        cleanPersistenceDir();
+
+        clearGridToStringClassCache();
+
+        GridQueryProcessor.idxCls = null;
+
+        super.afterTest();
+    }
+
+    /** */
+    @Test
+    public void testCorruptedTree() throws Exception {
+        IgniteEx srv = startGrid(0);
+        IgniteEx normalNode = startGrid(1);
+
+        normalNode.cluster().state(ClusterState.ACTIVE);
+
+        // Create SQL cache
+        IgniteCache<Integer, Integer> cache = srv.getOrCreateCache(DEFAULT_CACHE_NAME);
+
+        String tableName1 = "test1";
+        String tableName2 = "test2";
+
+        String qry = "create table %s (col1 varchar primary key, col2 varchar, col3 varchar, col4 varchar) with " +
+            "\"BACKUPS=1\"";
+
+        cache.query(new SqlFieldsQuery(String.format(qry, tableName1)));
+        cache.query(new SqlFieldsQuery(String.format(qry, tableName2)));
+
+        // Create indexes
+        cache.query(new SqlFieldsQuery("create index " + FAIL_IDX_1 + " on test1(col2) INLINE_SIZE 0"));
+        cache.query(new SqlFieldsQuery("create index " + OK_IDX + " on test1(col3) INLINE_SIZE 0"));
+        cache.query(new SqlFieldsQuery("create index " + FAIL_IDX_2 + " on test1(col4) INLINE_SIZE 0"));
+        cache.query(new SqlFieldsQuery("create index " + FAIL_IDX_3 + " on test2(col2) INLINE_SIZE 0"));
+
+        for (int i = 0; i < 100; i++) {
+            int counter = i;
+
+            String value = "test" + i;
+
+            String query = "insert into %s(col1, col2, col3, col4) values (?1, ?2, ?3, ?4)";
+
+            Stream.of(tableName1, tableName2).map(tableName -> new SqlFieldsQuery(String.format(query, tableName))
+                .setArgs(String.valueOf(counter), value, value, value)).forEach(cache::query);
+        }
+
+        IgniteH2Indexing indexing = (IgniteH2Indexing) srv.context().query().getIndexing();
+
+        String cache1Name = "SQL_PUBLIC_TEST1";
+        String cache2Name = "SQL_PUBLIC_TEST2";
+
+        corruptIndex(srv, indexing, cache1Name, FAIL_IDX_1);
+        corruptIndex(srv, indexing, cache1Name, FAIL_IDX_2);
+        corruptIndex(srv, indexing, cache2Name, FAIL_IDX_3);
+
+        // Check that index was indeed corrupted
+        checkIndexCorruption(srv.context().maintenanceRegistry(), cache, tableName1, asList("col2", "col4"));
+        checkIndexCorruption(srv.context().maintenanceRegistry(), cache, tableName2, singletonList("col2"));
+
+        stopGrid(0);
+
+        // This time node should start in maintenance mode and first index should be rebuilt
+        srv = startGrid(0);
+
+        assertTrue(srv.context().maintenanceRegistry().isMaintenanceMode());
+
```

Comment:
Please check that 3 DurableBackgroundCleanupIndexTreeTaskV2 have been created for each index.